poly/joint.go

```diff
@@ -0,0 +1,286 @@
+	if err == nil {
+		r.Dealers = append(r.Dealers, dealer)
+	}
+	return resp, err
+}
+
+// ProduceSharedSecret will generate the sharedsecret relative to this receiver
+// it will throw an error if something is wrong such as not enough Dealers received
+func (r *Receiver) ProduceSharedSecret() (*SharedSecret, error) {
+	if len(r.Dealers) < 1 {
+		return nil, errors.New("Receiver has 0 Dealers in its data.Can't produce SharedSecret.")
+	}
+	pub := new(PubPoly)
+	//pub.InitNull(r.info.Suite, r.info.T, r.Dealers[0].Promise.PubPoly().GetB())
+	pub.InitNull(r.info.Suite, r.info.T, r.info.Suite.Point().Base())
+	share := r.info.Suite.Secret()
+	goodShare := 0
+	for index, _ := range r.Dealers {
+		// Only need T shares
+		if goodShare >= r.info.T {
+			break
+		}
+		// Compute secret shares of the shared secret = sum of the respectives shares of peer i
+		// For peer i , s = SUM fj(i)
+		// TODO WARNING : THIS IS TEST IMPLEMENTATION
+		// In reality we should receive a NEW state struct from the dealer which is Certified so
+		// we can call RevealShare
+		// In testing we don't care about malicous yet so we just create one here
+		state := new(State).Init(*r.Dealers[index].Promise)
+		s, e := state.RevealShare(r.index, r.Key)
+		//s, e := r.Dealers[index].State.RevealShare(r.index, r.Key)
+		if e != nil {
+			//TODO error handling function not implemented right now. Only used for testing / comparison.
+			// We must be able to tell which share failed and to implement the broadcast of that error to others receiver
+			// so they reconstruct the private polynomial of the malicious dealer and set their share themself
+			return nil, errors.New(fmt.Sprintf("Receiver %d could not reveal its share from Dealer %d promise : %v", r.index, index, e))
+		}
+		share.Add(share, s)
+
+		// Compute shared public polynomial = SUM of indiviual public polynomials
+		pub.Add(pub, r.Dealers[index].Promise.PubPoly())
+
+		goodShare += 1
+	}
+
+	if goodShare < r.info.T {
+		return nil, errors.New("Not enough shares received by the Receiver to construct its own share of the shared secret")
+	}
+
+	if val := pub.Check(r.index, share); val == false {
+		return nil, errors.New("Receiver's secret share of the shared secret could not be checked against the shared polynomial")
+	}
+
+	return &SharedSecret{
+		Pub:   pub,
+		Share: &share,
+		Index: r.index,
+	}, nil
+}
+
+// MARSHALLING side
+
+// PolyInfo marshalling :
+// It is important to note that it does not Marshal / Unmarshal
+// the SUITE used. The suite is generated by default in constants.go
+// If you want to change to another suite, just change SUITE = ....
+func (p *PolyInfo) MarshalSize() int {
+	return util.IntSize(p.N) +
+		util.IntSize(p.R) +
+		util.IntSize(p.T)
+}
+
+func (p *PolyInfo) Equal(p2 PolyInfo) bool {
+	return p.N == p2.N && p.R == p2.R && p.T == p2.T && p.Suite.String() == p2.Suite.String()
+}
+
+func (p *PolyInfo) MarshalBinary() ([]byte, error) {
```

Comment:
Hi,

ah nice it seems you already figured out the problem regarding abstract.Read/Write. I also looked into it a little bit earlier today but then ran out of time to post my findings. In any case, I came to the same conclusion that abstract.Read/Write seems not to be working with basic types. See examples below. 

Furthermore, it seems to be a known approach when using the encoding/binary package that you have to Write/Read all the elements of a struct step-by-step. See left side of: https://bochs.info/p/cxvm9

Cheers,
Philipp

```go
package main

import (
    "bytes"
    "fmt"
    "reflect"
    "github.com/dedis/crypto/abstract"
    "github.com/dedis/crypto/edwards"
)

var suite = edwards.NewAES128SHA256Ed25519(true)

type Struct01 struct {
    Secret abstract.Secret
    Point abstract.Point
}

// this test passes since the to-be-(un)packed struct has only abstract.Secret
// and abstract.Point elements
func test01() {

    rand := suite.Cipher([]byte("example"))
    x := suite.Secret().Pick(rand)
    y := suite.Point().Mul(nil, x)

    cs := Struct01{
        Secret: x,
        Point: y,
    }

    buf := new(bytes.Buffer)
    err := abstract.Write(buf, cs, suite)
    if err != nil {
        fmt.Printf("Error Writing : %v\n", err)
    }

    buf2 := bytes.NewBuffer(buf.Bytes())
    cs2 := Struct01{}
    err = abstract.Read(buf2, &cs2, suite)
    if err != nil {
        fmt.Printf("Error Reading : %v\n", err)
    }

    fmt.Printf("Test01: ")
    if reflect.DeepEqual(cs,cs2) {
        fmt.Print("Hurray!\n")
    } else {
        fmt.Printf("Nope!\n")
        fmt.Println(cs)
        fmt.Println(cs2)
    }
}


type Struct02 struct {
    Index int32
    Arr []byte
    Secret abstract.Secret
    Point abstract.Point
}

// this test crashes since the to-be-(un)packed struct includes elements with
// types other than abstract.Secret and abstract.Point
func test02() {

    rand := suite.Cipher([]byte("example"))
    x := suite.Secret().Pick(rand)
    y := suite.Point().Mul(nil, x)

    cs := Struct02{
        Index: 4, // it seems this value is not read at all by abstract.Read
        Arr: []byte{5,6}, // during abstract.Read the two bytes are merged into the Secret below which in turn corrupts reading of Secret and Point
        Secret: x,
        Point: y,
    }

    buf := new(bytes.Buffer)
    err := abstract.Write(buf, cs, suite)
    if err != nil {
        fmt.Printf("Error Writing : %v\n", err)
    }

    buf2 := bytes.NewBuffer(buf.Bytes())
    cs2 := Struct02{}
    err = abstract.Read(buf2, &cs2, suite)
    if err != nil {
        fmt.Printf("Error Reading : %v\n", err)
    }

    fmt.Printf("Test02: ")
    if reflect.DeepEqual(cs,cs2) {
        fmt.Printf("Hurray!\n")
    } else {
        fmt.Printf("Nope!\n")
        fmt.Println(cs)
        fmt.Println(cs2)
    }
}


func main() {
    test01()
    test02()
}
```